### DISCUSSION TITLE
Is there a way to  override resource models?

### DISCUSSION THREAD

**Discussion by csl-schlewitz** [OpenMage/magento-lts#4001] [Questions & Answers] [answered]
I tried to override the resource model **Mage_Reports_Model_Resource_Product_Collection**
I have to modify the function **addOrderedQty**

I used the default approach to override the model in the config.xml

```xml
<global>
<models>
          <tenretni_reports>
                <class>Tenretni_Reports_Model</class>
          </tenretni_reports>
          <reports>        
            <rewrite>
                <resource_product_collection>Tenretni_Reports_Model_Resource_Product_Collection</resource_product_collection>                   
            </rewrite>
          </reports>
</models>
</global>
```

after exhausting all other options / solutions i have knowledge about , i just wonder if its at all possible to override
resource models this way or  do i need a different  approach...?

( i know that i can use a mirrored file in _local/Mage_  to override the whole file , but i wanted to avoid this approach. )

**Comment by fballiano** [comment]
what if you override both the model and the resourcemodel?

**Reply by csl-schlewitz** [reply]
there is no Model in that case... it addressed over the grid system.

Mage_Adminhtml_Block_Report_Product_Sold_Grid

```
protected function _prepareCollection()
    {
        parent::_prepareCollection();
        $this->getCollection()
            ->initReport('reports/product_sold_collection');
        return $this;
    }
```

into

Mage_Reports_Model_Report

```
public function initCollection($modelClass)
    {
        $this->_reportModel = Mage::getResourceModel($modelClass);

        return $this;
    }
```

I also tried to override **Mage_Reports_Model_Resource_Product_Sold_Collection** that is extending....
But there no reaction to that all.

I was thinking about overriding the _prepareCollection from the Grid-Block.... But thats a lot of work and and i am not sure how the report module from the core might react to that

**Comment by luigifab** [comment, marked as answer]
To override a resource model, here `Mage_Reports_Model_Resource_Product_Collection`, you must use `reports_resource` and not `reports`:

https://github.com/OpenMage/magento-lts/blob/e22853875f197abc3413797f75e59d00529ba9c2/app/code/core/Mage/Reports/etc/config.xml#L24-L30

**Reply by csl-schlewitz** [reply]
Thanks for pointing that out.
Resource-Models having their own "namespace", i failed to recognize that.

But The override is still not working.

i allready did try handling the rewrite unter "reports_resource"
With this insight however i was wondering if i maybe also have to mirror the namespace in my module...

so my model setup would result to this

```xml
<?xml version="1.0"?>
<config>
    <modules>
        <Tenretni_Reports>
            <version>1.0.0</version>
        </Tenretni_Reports>
    </modules>

    <global>
        <blocks>
            <tenretni_reports>
                <class>Tenretni_Reports_Block</class>
            </tenretni_reports>
        </blocks>

        <models>
            <tenretni_reports>
                <class>Tenretni_Reports_Model</class>
                <resourceModel>tenretni_reports_resource</resourceModel>
            </tenretni_reports>
            <tenretni_reports_resource>
                <class>Tenretni_Reports_Model_Resource</class>
            </tenretni_reports_resource>

            <reports_resource>
                <rewrite>
                    <product_collection>Tenretni_Reports_Model_Resource_Product_Collection</product_collection>
                </rewrite>
            </reports_resource>
        </models>

        <helpers>
            <tenretni_reports>
                <class>Tenretni_Reports_Helper</class>
            </tenretni_reports>
        </helpers>

        <resources>
            <tenretni_reports_setup>
                <setup>
                    <module>Tenretni_Reports</module>
                </setup>
            </tenretni_reports_setup>
        </resources>
    </global>
</config>
```

But ultimately the override is still not working.

I was thinking that my module is maybe faulty  but i could call a test funtion in the modules helper...
So i am pretty sure the module in itself should be fine.

**Reply by kiatng** [reply]
I tested your config in my module, I can confirm it is working.
```xml
        <models>
            <!-- ... -->
            <reports_resource>
                <rewrite>
                    <product_collection>Celera_Crib_Model_Resource_Product_Collection</product_collection>
                </rewrite>
            </reports_resource>
        </models>
```

```php
<?php
class Celera_Crib_Model_Resource_Product_Collection extends Mage_Reports_Model_Resource_Product_Collection
{
}
```

Test script in a controller
```php
    public function indexAction()
    {
        $class = Mage::getResourceModel('reports/product_collection');
        $className = get_class($class);
        $this->getResponse()->setHeader('Content-Type', 'text/html')->setBody($className);
    }
```

Output in browser:
Celera_Crib_Model_Resource_Product_Collection

**Reply by csl-schlewitz** [reply]
I see, i did another check and your are both right.

The resource Model that was loaded in my case
```
class Mage_Reports_Model_Resource_Product_Sold_Collection extends Mage_Reports_Model_Resource_Product_Collection
{
```
Thats way the Override was not working....
I also tried to override the product_sold_collection before my inquiry , but now it is working.

So i guess what happend on my hand was that i never did try to override the "product_sold_collection" under the correct namespace

i reverted my code back to 

```
<models>
            <tenretni_reports>
                <class>Tenretni_Reports_Model</class>
            </tenretni_reports>

            <reports_resource>
                <rewrite>
                    <product_collection>Tenretni_Reports_Model_Resource_Product_Collection</product_collection>
                    <product_sold_collection>Tenretni_Reports_Model_Resource_Product_Sold_Collection</product_sold_collection>
                </rewrite>
            </reports_resource>
</models>
```
And this is working now.
Thanks for the Help
